File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        inDegree = [0 for _ in range(numCourses)]
        queue = []
        adj = [[] for _ in range(numCourses)]
        for arr in prerequisites:
            adj[arr[0]].append(arr[1])
        for arr in adj:
            for i in arr:
                inDegree[i]+=1
        for i in range(numCourses):
            if inDegree[i] == 0:
                queue.append(i)
        while(queue):
            curr = queue.pop(0)
            numCourses-=1
            for nextNode in adj[curr]:
                inDegree[nextNode]-=1
                if inDegree[nextNode] == 0:
                    queue.append(nextNode)
        if numCourses == 0:
            return True
        return False

File: 207-course-schedule/course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        for(int i = 0;i<numCourses;i++){
            graph.add(new ArrayList<>());
        }
        int count = 0;
        int [] inDegrees = new int[numCourses];
        for(int i = 0;i<prerequisites.length;i++){
            graph.get(prerequisites[i][1]).add(prerequisites[i][0]);
            inDegrees[prerequisites[i][0]]+=1;
        }
        Stack<Integer> stack = new Stack();
        for(int i = 0;i<inDegrees.length;i++){
            if(inDegrees[i] == 0){
                stack.push(i);
            }
        }
        while(!stack.isEmpty()){
            int currentCourse = stack.pop();
            count+=1;
            for(int nextCourse : graph.get(currentCourse)){
                inDegrees[nextCourse]-=1;
                if(inDegrees[nextCourse] == 0){
                    stack.push(nextCourse);
                }
            }
        }
        if(count == numCourses){
            return true;
        }
        return false;
    }
}